mobile_verifier/src/heartbeats/mod.rs

```diff
@@ -256,17 +256,54 @@ impl HeartbeatReward {
         self.reward_weight
     }
 
-    pub fn validated<'a>(
+    pub async fn validated<'a>(
         exec: impl sqlx::PgExecutor<'a> + Copy + 'a,
         epoch: &'a Range<DateTime<Utc>>,
         max_distance_to_asserted: u32,
-    ) -> impl Stream<Item = Result<HeartbeatReward, sqlx::Error>> + 'a {
-        sqlx::query_as::<_, HeartbeatRow>(include_str!("valid_heartbeats.sql"))
-            .bind(epoch.start)
-            .bind(epoch.end)
-            .bind(MINIMUM_HEARTBEAT_COUNT)
-            .fetch(exec)
-            .map_ok(move |row| Self::from_heartbeat_row(row, max_distance_to_asserted))
+    ) -> anyhow::Result<impl Stream<Item = HeartbeatReward> + 'a> {
+        let heartbeat_rows =
+            sqlx::query_as::<_, HeartbeatRow>(include_str!("valid_heartbeats.sql"))
+                .bind(epoch.start)
+                .bind(epoch.end)
+                .bind(MINIMUM_HEARTBEAT_COUNT)
+                .fetch(exec)
+                .try_fold(
+                    HashMap::<(PublicKeyBinary, Option<String>), Vec<HeartbeatRow>>::new(),
+                    |mut map, row| async move {
+                        map.entry((row.hotspot_key.clone(), row.cbsd_id.clone()))
+                            .or_default()
+                            .push(row);
+
+                        Ok(map)
+                    },
+                )
+                .await?;
+
+        Ok(
+            futures::stream::iter(heartbeat_rows).map(move |((hotspot_key, cbsd_id), rows)| {
+                let first = rows.first().unwrap();
+                let average_location_trust_score = rows
+                    .iter()
+                    .map(|row| {
+                        row.cell_type.location_weight(
+                            row.location_validation_timestamp,
+                            row.distance_to_asserted,
+                            max_distance_to_asserted,
+                        )
+                    })
+                    .sum::<Decimal>()
+                    / Decimal::new(rows.len() as i64, 0);
+
+                HeartbeatReward {
+                    hotspot_key,
+                    cell_type: first.cell_type,
```

Comment:
guessing this is the thing you weren't happy about? feels like this is probably not the right abstraction if we're arbitrarily peeling off the first value of the list to grab these field values but understand now's not the time to fix it. probably something where the `HeartbeatReward` is just an envelope struct that contains the contents of the `HeartbeatRow` but then adds in the newly computed `reward_weight`

Reply:
I don't like this, but it works and the all the fields we are reading from the first row are the same in all rows.